src/mame/sinclair/specnext_sprites.cpp

```diff
@@ -0,0 +1,282 @@
+
+static const gfx_layout gfx_16x16x8 =
+{
+    16, 16, RGN_FRAC(1, 1), 8,
+    { STEP8(0, 1) },
+    { STEP16(0, 8) },
+    { STEP16(0, 8 * 16) },
+    8 * 16 * 16
+};
+
+static const gfx_layout gfx_16x16x4 =
+{
+    16, 16, RGN_FRAC(1, 1), 4,
+    { STEP4(0, 1) },
+    { STEP16(0, 4) },
+    { STEP16(0, 4 * 16) },
+    4 * 16 * 16
+};
+
+static const gfx_layout gfx_16x16x8_r =
+{
+    16, 16, RGN_FRAC(1, 1), 8,
+    { STEP8(0, 1) },
+    { STEP16(8 * 16 * 15, -8 * 16) },
+    { STEP16(0, 8) },
+    8 * 16 * 16
+};
+
+static const gfx_layout gfx_16x16x4_r =
+{
+    16, 16, RGN_FRAC(1, 1), 4,
+    { STEP4(0, 1) },
+    { STEP16(4 * 16 * 15, -4 * 16) },
+    { STEP16(0, 4) },
+    4 * 16 * 16
+};
+
+static GFXDECODE_START( gfx_sprites )
+    GFXDECODE_DEVICE_RAM( "pattern_ram", 0, gfx_16x16x8, 0x600, 256 )
+    GFXDECODE_DEVICE_RAM( "pattern_ram", 0, gfx_16x16x4, 0x600, 16 )
+    GFXDECODE_DEVICE_RAM( "pattern_ram", 0, gfx_16x16x8_r, 0x600, 256 )
+    GFXDECODE_DEVICE_RAM( "pattern_ram", 0, gfx_16x16x4_r, 0x600, 16 )
+GFXDECODE_END
+
+specnext_sprites_device::specnext_sprites_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock)
+	: device_t(mconfig, SPECNEXT_SPRITES, tag, owner, clock)
+    , device_gfx_interface(mconfig, *this, gfx_sprites, ":palette")
+    , m_sprite_pattern_ram(*this, "pattern_ram", 64 * 256, ENDIANNESS_LITTLE)
+    , m_sprite_attr_ram(*this, "attr_ram", 128 * 8, ENDIANNESS_LITTLE)
+{
+}
+
+void specnext_sprites_device::draw(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect)
+{
+    if (m_sprites_cache.empty()) update_sprites_cache();
+    const rectangle clipped = m_clip_window & cliprect;
+
+    for (auto i = 0; i < m_sprites_cache.size(); i++ ) {
+        const u8 x = m_zero_on_top ? i : (m_sprites_cache.size() - 1) - i;
+        const sprite_data* spr = &m_sprites_cache[x];
+
+        gfx((spr->rotate << 1) | spr->h)->zoom_transpen(bitmap, clipped
+            , spr->pattern, spr->paloff
+            , spr->xmirror, spr->ymirror
+            , (spr->x & 0x1ff) + m_offset.first, (spr->y & 0x1ff) + m_offset.second
+            , 0x10000 << spr->xscale, 0x10000 << spr->yscale
+            , m_transp_colour);
+    }
+}
+
+void specnext_sprites_device::update_sprites_cache()
+{
+    m_sprites_cache.reserve(128);
+
+    sprite_data *anchor;
+    bool anchor_vis = false;
+    u8* sprite_attr = m_sprite_attr_ram;
+    for (auto i = 0; i < 128; i++, sprite_attr += 8)
+    {
+        const bool spr_relative = BIT(sprite_attr[3], 6) && ((sprite_attr[4] >> 6) == 0b01);
+        bool is_visible = BIT(sprite_attr[3], 7);
+
+        if (spr_relative)
+            is_visible &= anchor_vis;
+        else
+            anchor_vis = is_visible;
+
+        if (is_visible)
+        {
+            u8 spr_cur_attr[5] = {};
+            if (!spr_relative) // Anchor
+            {
+                memcpy(spr_cur_attr, sprite_attr, 5);
+                anchor = &m_sprites_cache.back() + 1;
```

Comment:
Taking a pointer to a member of a vector that doesn’t exist is pretty bad form.  Since you appear to always have 128 sprites, why are you using a vector in the first place?  You could just use an array, or a dynamically allocated `std::array`, or something.

Reply:
Not always. We have 128 at max, but only anchors and visible relative are stored in the cache.
With array I need to introduce extra state variable for size.